File: basic/week(5)/higher_order_fun.py

from functools import reduce
numbers = [1, 2, 3, 4]


squared = list(map(lambda x: x * x, numbers))
print(squared) 

 
evens = list(filter(lambda x: x % 2 == 0, numbers))
print(evens) 


sum = reduce(lambda acc, x: acc + x, numbers, 0)
print(sum)  

sorted_numbers = sorted(numbers, key=lambda x: x)
print(sorted_numbers) 



def custom_map(callback, numbers):
    res = []
    for i in range(len(numbers)):
        res.append(callback(numbers[i], i, numbers)) 
    return res   
def custom_filter(callback, numbers):
    res = []
    for i in range(len(numbers)):
        if callback(numbers[i], i, numbers): 
            res.append(numbers[i]) 
    return res      
    
def custom_reduce(callback, numbers, initial_value=None):
    res = initial_value if initial_value is not None else numbers[0]  
    start_index = 0 if initial_value is not None else 1  

    for i in range(start_index, len(numbers)):
        res = callback(res, numbers[i], i, numbers) 
    return res



squared = list(custom_map(lambda x: x * x, numbers))
print(squared) 

 
evens = list(custom_filter(lambda x: x % 2 == 0, numbers))
print(evens) 


sum = custom_reduce(lambda acc, x: acc + x, numbers, 0)
print(sum)  

File: basic/week(5)/higher_order_fun.js
const numbers = [1, 2, 3, 4];

const squared = numbers.map(x => x * x);
console.log(squared); 



const evens = numbers.filter(x => x % 2 === 0);
console.log(evens); 


const sum = numbers.reduce((acc, x) => acc + x, 0);
console.log(sum); 


const sorted = numbers.sort((a, b) => a - b);
console.log(sorted);

function customMap(arr,callback){
    let result=[];
    for(let a in arr){
        result.push(callback(arr[a],a,arr));
    }
    return result;
}

const customsquared = customMap(numbers, x => x * x);
console.log("Squared:", customsquared);


function customReduce(arr,callback,initalval){
    let acc=initalval!==undefined?initalval:arr[0];
    let startindex=initalval!==undefined?0:1;
    for(let i=startindex;i<arr.length;i++){
        acc=callback(acc,arr[i],i,arr)
    }
    return acc;
}
customReduce(numbers,(acc,x)=>acc+x,0);


function customFilter(arr, callback) {
    const result = [];
    for (let i = 0; i < arr.length; i++) {
        if (callback(arr[i], i, arr)) {
            result.push(arr[i]);
        }
    }
    return result;
}


const cus_evens = customFilter(numbers, x => x % 2 === 0);
console.log(cus_evens)
